brave/brave-core
# [ads] Transition text classification v2/v1 models from JSON to FlatBuffers

Current PR does the following:
* Adds support of new v2 (neural) FlatBuffers text classification model
* Adds support of old v1 (linear) text classification model converted to FlatBuffers (the previous format was JSON)
* Tries to load v2 (neural) text classification model first
* Fallbacks to the old v1 (linear) text classification model, if the v2 (neural) model failed


Resolves https://github.com/brave/brave-browser/issues/33336



## Submitter Checklist:

  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:

### Test case 1
* Set English language system locale
* Run the browser with a fresh profile
* Join Brave Rewards
* Make sure that Brave Ads resource for EN language was downloaded on brave://components
* Make sure that the v2 (neural) text classification resource is loaded and initialized. There should be a line in the brave rewards log: `Successfully initialized awdrtoxdlcdyjmckgmtqptnhdtlxgblp neural text classification resource`
* Visit a web page. Make sure that the page is classified. For example visiting `https://bitcoin.org/en/` gives the following line in brave rewards log: `Classified text with the top segment as crypto-crypto`
* Make sure that notification ads can be served.

### Test case 2
* Set non-English language system locale
* Run the browser with a fresh profile
* Join Brave Rewards
* Make sure that Brave Ads resource for language was downloaded on brave://components
* Make sure that the v1 (linear) text classification resource is loaded and initialized. There should be a line in the brave rewards log: `Successfully initialized awdrtoxdlcdyjmckgmtqptnhdtlxgblp linear text classification resource`
* Visit a web page. Make sure that the page is classified.
* Make sure that notification ads can be served.

### Test case 3
* Set English language system locale
* Run the browser with a fresh profile
* Join Brave Rewards
* Make sure that Brave Ads resource for EN language was downloaded on brave://components
* Exit Brave Browser
* Open Brave Ads resource component directory at `Brave-Browser/kglppciafpehhjdhgefiljcmemakfjhk/{version}`
* Delete `svymudlfwueobbiprqazdzkdeslimohu` file
* Create `svymudlfwueobbiprqazdzkdeslimohu` file with some random text
* Run the browser
* Make sure that the text classification resource isn't loaded. There should be a line in the brave rewards log: `Failed to initialize awdrtoxdlcdyjmckgmtqptnhdtlxgblp text classification resource (Failed to load flatbuffers text classification pipeline)`
* Make sure that notification ads can be served.
